jcl/src/java.base/share/classes/java/lang/invoke/MethodHandles.java

```diff
@@ -489,6 +489,13 @@ private void checkClassAccess(Class<?> targetClass) throws IllegalAccessExceptio
 					return;
 				}
 			} else {
+				/* target class should always be accessible to the lookup class when they are the same class,
+				 * as long as there is no restriction in the access mode.
+				 */
+				if ((NO_ACCESS != accessMode) && (accessClass == targetClass)) {
```

Comment:
Can we check for NO_ACCESS at the start of the method?  If the accessMode is NO_ACCESS, then not even public members should be accessible...

The same goes for the `accessClass == targetClass` check - if its the same class, then none of the other checks are required so let's do that second.

```
if (NO_ACCESS != accessMode) {
  if (accessClass == targetClass) {
     return;
  }
  if (Modifier.isPublic(modifiers)) {.....

}

Reply:
Can you also delete the `isProtected` check as that should never occur with a class?

Reply:
1) Can we check for NO_ACCESS at the start of the method?  <----- Agreed and fixed.
2) Can you also delete the isProtected check ? <---- Already removed and will verify with the existing JSR292 test case and JCK tests to ensure everything works good.

Reply:
Can you also delete the isProtected check ? 

This is incorrect for a nested/inner protected class as it set up the modifier with memberAccessFlags (protected) instead of the modifier of the class itself (public)

ClassFileOracle::walkAttributes()
{
...
		case CFR_ATTRIBUTE_InnerClasses: {
..
if (innerClassUTF8 == thisClassUTF8) {
					_isInnerClass = true;
					_memberAccessFlags = entry->innerClassAccessFlags; <---------------
...


	VMINLINE VM_BytecodeAction
	inlClassGetModifiersImpl(REGISTER_ARGS_LIST)
	{
	...
			U_32 modifiers = 0;
		if (J9_ARE_NO_BITS_SET(romClass->extraModifiers, J9AccClassInnerClass)) {
			/* Not an inner class - use the modifiers field */
			modifiers = romClass->modifiers;
		} else {
			/* Use memberAccessFlags if the receiver is an inner class */
			modifiers = romClass->memberAccessFlags; <------------------------ protected
		}

Fast_java_lang_Class_getModifiersImpl(J9VMThread *currentThread, j9object_t receiverObject)
{
...
	U_32 modifiers = 0;
	if (J9_ARE_NO_BITS_SET(romClass->extraModifiers, J9AccClassInnerClass)) {
		/* Not an inner class - use the modifiers field */
		modifiers = romClass->modifiers;
	} else {
		/* Use memberAccessFlags if the receiver is an inner class */
		modifiers = romClass->memberAccessFlags; <------------------------- protected
	}

------------------
The failure was captured by one of our JSR292 tests in test_FindNestedClass_ProtectedAccess_DiffPackage as follows:
java.lang.IllegalAccessException: 'com.ibm.j9.jsr292.LookupAPITests_FindClass' no access to: 'com.ibm.j9.jsr292.helpers.Helper_LookupAPI_OtherPackage$ProtectedNestedClass'
	at java.lang.invoke.MethodHandles$Lookup.checkClassAccess(java.base@9-internal/MethodHandles.java:444)
	at java.lang.invoke.MethodHandles$Lookup.checkAccess(java.base@9-internal/MethodHandles.java:334)
	at java.lang.invoke.MethodHandles$Lookup.accessClass(java.base@9-internal/MethodHandles.java:1432)
	at java.lang.invoke.MethodHandles$Lookup.findClass(java.base@9-internal/MethodHandles.java:1415)
	at com.ibm.j9.jsr292.LookupAPITests_FindClass.test_FindNestedClass_ProtectedAccess_DiffPackage(LookupAPITests_FindClass.java:203)

--------------------------
The idea is that a protected nested/inner class (treated as public) should be accessible to other classes.  So the check for Modifier.isProtected(modifiers) should be kept there in such case.

Correct me if anything wrong.

Reply:
Does that test pass on OpenJDK?

Reply:
Yes, they all passed on Open JDK. Please see the results at https://github.com/eclipse/openj9/issues/636